scripts/hud/comparisons.ts

```diff
@@ -1,140 +1,325 @@
+			{
+				class: 'hud-splits__diff diff',
+				text: '{s:diff_sign}{g:time:diff}'
+			}
+		)
+	}));
+
+	// - When timer progression changes:
+	//   - Generate progression split, set bottom-most split to it
+	//   - Push all other splits up by setting them to the previous split's properties
+	// - When timer state changes,
+	//   - To primed: Clear all splits
+	//   - To running: Clear all splits
+	//   - To disabled: Clear all splits
+	//   - To finished: Generate final split, set bottom-most split to it
+	// - When comparison changes:
+	//   - Update all splits to use new comparison
+	// - When timer replaced:
+	//   - Regenerate everything. During seeking we'll do this a *lot*, but this
+	//     design is very performant and can handle it. If we ever need to speed it
+	//     up, could add an event specifically for seeking.
 	constructor() {
-		$.RegisterForUnhandledEvent('ComparisonRunUpdated', () => this.onComparisonRunUpdated());
-		$.RegisterForUnhandledEvent('OnObservedTimerStateChange', () => this.updateComparisons());
-		$.RegisterForUnhandledEvent('OnObservedTimerCheckpointProgressed', () => this.updateComparisons());
-		$.RegisterForUnhandledEvent('OnObservedTimerReplaced', () => this.updateComparisons());
-	}
+		$.RegisterForUnhandledEvent('OnObservedTimerStateChange', () => {
+			const { state, runTime, segmentsCount, segmentCheckpointsCount } =
+				MomentumTimerAPI.GetObservedTimerStatus();
 
-	onComparisonRunUpdated() {
-		this.comparison = RunComparisonsAPI.GetComparisonRun();
+			if (state === Timer.TimerState.FINISHED && (segmentsCount > 1 || segmentCheckpointsCount > 1)) {
+				const splits = MomentumTimerAPI.GetObservedTimerRunSplits();
 
-		$.Msg(`Got comparison run ${this.comparison?.filePath ?? 'NONE'}`);
+				this.updateLatestSplit(
+					Timer.generateFinishSplit(
+						splits,
+						this.comparison?.runSplits ?? null,
+						runTime,
+						this.comparison?.runTime ?? 0,
+						segmentsCount,
+						segmentCheckpointsCount
+					)
+				);
+			} else {
+				this.clearSplits();
+			}
+		});
 
-		if (this.comparison) {
-			$.Msg(`jumps: ${this.comparison.runSplits.trackStats.jumps}`);
-		}
+		$.RegisterForUnhandledEvent('ComparisonRunUpdated', () => {
+			const { state } = MomentumTimerAPI.GetObservedTimerStatus();
+			// If timer just finished and we're not watching a replay, don't update - otherwise
+			// comparison will get set to the run you just finished when you PB which we
+			// obviously don't want to happen.
+			if (state === Timer.TimerState.FINISHED && this.controlledReplayID === null) {
+				return;
+			}
+
+			this.comparison = RunComparisonsAPI.GetComparisonRun();
+			this.recomputeComparisons();
+		});
+
+		$.RegisterForUnhandledEvent('OnObservedTimerCheckpointProgressed', () => {
+			const { majorNum, minorNum, segmentsCount, segmentCheckpointsCount } =
+				MomentumTimerAPI.GetObservedTimerStatus();
+
+			const splits = MomentumTimerAPI.GetObservedTimerRunSplits();
+
+			this.updateLatestSplit(
```

Comment:
I think the splits HUD needs to not do any comparison stuff if the comparison is for a different track.

In theory if your comparison run is a stage and you are running the Main track (or vice-versa) the comparisons logic could identify that and line up the comparison with the appropriate stage -- however to *actually* make that useful, what we are really going to want is the ability to do like a "auto set comparison to best of each stage" if you want to run the Main track but compare to individual stage times, so I think we will get to that later and then that would be a good time to do the stage comparison alignment thing.

Reply:
Yeah good spot, fixed in `438c0d5`